TT-Fabric Bringup Initial Check-in

aliuTT opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/13844
https://github.com/tenstorrent/tt-metal/issues/13845

### Problem description
Combined work for initial bringup of TT-FABRIC, by @aagarwalTT @ubcheema @aliuTT

### What's changed
In first check-in of TT-FABRIC, we have only added new components and tests for it. There should be no functional changes to existing runtime/FD/ops etc.

Still need to rebase and run CI, but from top of my head there are some known issues:

- ~~@ubcheema currently, I've added the eth chan noc encoding map to the `generated_bank_to_noc_coord_mapping.h`, which is removed on latest main. Need to figure out where to load this. Commit `d9f211596` needs special attention.~~ resolved
- ~~we should add the passing tests to the respective CIs~~ resolved
- some commented out code, can be removed

### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) tests passes
- [ ] New/Existing tests provide coverage for changes

aliuTT committed:
#13844: add initial bringup of routing table, control plane, mesh graph

ubcheema committed:
#13845: add tt-fabric router components

ubcheema committed:
Fix fabric worker bugs and integrate with routing tables

router bug fixes
add fabric atomic increment command, rearrange handling of trailing packet flush.
Add fabric interface header for interface structures and address map
Plug in routing table look up to determine next hop.

aliuTT committed:
Add routing table support for t3k

aliuTT committed:
ControlPlane: add getter apis

physical to mesh device id conversions
fabric route

ubcheema committed:
Update tests

ubcheema committed:
Fix perf bug with frequent context switches. Add explicit terminate signal for routers.

aagarwalTT committed:
Add n300 support and fix tests

ubcheema commented:
@Allan, the eth channel to noc x/y mapping is fixed. We can add the mapping look up table to tt_faric header, instead of generating it on the fly.
What do you think?

ubcheema commented on tt_fabric/mesh_graph_descriptors/t3k_mesh_graph_descriptor.yaml:
```diff
@@ -0,0 +1,27 @@
+  id: 0,
+  board: T3K,
+  topology: [1, 1],
+  host_mapping: [[sjc-lab-t3005]]}
```
Is this descriptor tied to this specific host?

ubcheema commented on tt_fabric/mesh_graph_descriptors/tg_mesh_graph_descriptor.yaml:
```diff
@@ -0,0 +1,66 @@
+  id: 0,
+  board: N150Gateway,
+  topology: [1, 1],
+  host_mapping: [[aus-glx-11]]},
```
Is this tg descriptor tied to aus-glx-11?
I suppose control plane is not currently using this field, so it does not matter...

aliuTT replied:
Yea, I can remove the names of the hosts.

ubcheema committed:
#0: fix typo bug in address calculation.
#0: limit max pull block size to half of fvc buffer

tt-rkim commented on tt_fabric/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+target_link_libraries(tt_fabric PRIVATE yaml-cpp::yaml-cpp)
+target_link_libraries(
+    tt_fabric
+    PUBLIC
```
Why does this need to be public?

tt-rkim replied:
Do you want people to be able to use metal via fabric? Seems like reaching around to me

tt-rkim commented on tt_fabric/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+
+target_include_directories(
+    tt_fabric
+    PUBLIC
```
Why does this need to be public? Too much pollution

aliuTT replied:
moved to private

tt-rkim commented on tt_fabric/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+    tt_fabric
+    PROPERTIES
+        INSTALL_RPATH
+            "${PROJECT_BINARY_DIR}/lib"
```
Please use `$ORIGIN` like we do for `_ttnn`

tt-rkim reviewed (dismissed):
First pass at CMake, but I'll definitely have stuff to say about the others

aliuTT commented:
> @allan, the eth channel to noc x/y mapping is fixed. We can add the mapping look up table to tt_faric header, instead of generating it on the fly. What do you think?

We can do that for now.

aagarwalTT committed:
test infra: support for B/W measurement, using multiple links, and minor fixes

aliuTT committed:
Rebase, PR feedback

- Remove host names
- Remove unnecessary loop in control plane routing table config
- Add explicit, use ttl reflection, clang-format
- Nicer connectivity populating, clang-format
- Fabric fixture cleanup

afuller-TT commented on tt_fabric/CMakeLists.txt:
```diff
@@ -0,0 +1,61 @@
+#add_library(tt_fabric ${TT_FABRIC_OBJECTS})
+
+add_library(tt_fabric)
+add_library(Metalium::Fabric ALIAS tt_fabric)
```
I started this nomenclature based on the assumption that Metal == Metalium.

Since recently finding a doc that spells out the (current) names for everything, I now come to understand that "tt-metal" repo is actually "TT-Metalium And Friends".

To reflect that, I think this would be more in line with intended names:

```suggestion
add_library(TT::Fabric ALIAS tt_fabric)
```

aliuTT replied:
I can make this change.

afuller-TT commented on tt_fabric/CMakeLists.txt:
```diff
@@ -0,0 +1,61 @@
+add_library(tt_fabric)
+add_library(Metalium::Fabric ALIAS tt_fabric)
+
+target_sources(
```
Thank you for modern and clean CMake 🙏 

FYI: the `${CMAKE_CURRENT_SOURCE_DIR}` is redundant and actually preferred to be omitted so that in the rare case that we we build a generated file, `${CMAKE_CURRENT_BINARY_DIR}` actually stands out as an exception rather than blends in to the noise.

afuller-TT commented on tt_fabric/CMakeLists.txt:
```diff
@@ -0,0 +1,61 @@
+target_include_directories(
+    tt_fabric
+    PUBLIC
+        ${PROJECT_SOURCE_DIR}
```
Please no.  I'm in the process of cleaning this up.  Nobody consuming TT-Fabric should know anything about anything outside the scope of TT-Fabric and we shouldn't enable it.  If they need TT-Metalium, then should ask TT-Metalium; if they need TT-NN, they should ask for TT-NN.

Do you remember what prompted this?  Or just a copy/paste of the status-quo elsewhere?

aliuTT replied:
Copy paste :) I've redone the CMakeList after rebasing to get the cleanup you've put in for existing ones. Is your cleanup PR likely to hit main this week? If not, can you paste the changes here so I can put into this PR?

afuller-TT replied:
The cleanup is a 800+ file PR: #16578 

We'll coordinate the final cleanup either in whichever PR lands later or in a fresh pass after both land.

afuller-TT commented on tt_fabric/CMakeLists.txt:
```diff
@@ -0,0 +1,61 @@
+    tt_fabric
+    PUBLIC
+        ${PROJECT_SOURCE_DIR}
+        ${CMAKE_CURRENT_SOURCE_DIR}
```
Is this the location of the TT-Fabric API for consumers?  There's no `api/` directory?  The target should only expose what's actually public and keep the private bits out of reach from consumers.

This is for the benefit of both the consumer (easy to see what's actually the API) and also for the maintainers (can YOLO internal files, but be more careful/meticulous with the exposed API).  Also tooling is available to more thoroughly test published headers.

aliuTT replied:
This is just an initial bringup of Fabric, all the major components of Fabric are getting checked in, but we haven't switched existing SW to use it yet (that is coming in the next month or so). We'll have a better idea of what APIs makes sense to expose as Fabric settles into the codebase more. I can remove the PUBLIC section of the includes, and add back a public API folder in the future.

afuller-TT replied:
It's 1000x easier to turn a private header public than to turn a public header private.  I suggest starting with nothing public, and then cherry-pick into public exposure on a case-by-case basis.  It also provides a natural gate to tidy up the header when it's promoted, especially if only half ought to be public.

afuller-TT commented on tt_fabric/CMakeLists.txt:
```diff
@@ -0,0 +1,61 @@
+        ${PROJECT_SOURCE_DIR}
+        ${CMAKE_CURRENT_SOURCE_DIR}
+    PRIVATE
+        ${UMD_HOME}
```
Shouldn't need this.  If you need UMD, then add a dependency on UMD.  I see line 24 has `umd::device`.  That should provide you what UMD deems its public API.  No need to manually set include dirs.

afuller-TT commented on tt_fabric/CMakeLists.txt:
```diff
@@ -0,0 +1,61 @@
+        ${CMAKE_CURRENT_SOURCE_DIR}
+    PRIVATE
+        ${UMD_HOME}
+        ${PROJECT_SOURCE_DIR}/tt_metal
```
This one's more questionable whether it's in a suitable shape.  Ideally `Metalium::Metal` on line 23 is providing you everything you need.  I'm currently going through to ensure that's the case.  Does TT-Fabric work if you drop the line?  Or are bits missing from Metalium::Metal?

aliuTT replied:
Dropped both UMD_HOME and tt_metal, in latest changes

afuller-TT commented on tt_fabric/CMakeLists.txt:
```diff
@@ -0,0 +1,61 @@
+        INSTALL_RPATH
+            "${PROJECT_BINARY_DIR}/lib"
+        ADDITIONAL_CLEAN_FILES
+            "${PROJECT_BINARY_DIR}/lib;${PROJECT_BINARY_DIR}/obj"
```
Are we manually placing files here (haven't looked that far yet)?  Or why do we need to tell CMake about files if we're not doing it behind CMake's back?

afuller-TT commented on tt_metal/hw/inc/wormhole/eth_l1_address_map.h:
```diff
@@ -7,6 +7,7 @@
 #include <cstdint>
 
 #include "noc/noc_parameters.h"
+#include "tt_fabric/hw/inc/routing_table.h"
```
Whoa!  We're creating a circular dependency from TT-Metalium reaching into TT-Fabric while TT-Fabric uses TT-Metalium?

Is this necessary/intentional?  If so, then it's time to think about extracting whatever this bit of logic is into a common dependency.  We really don't want circular deps.

aliuTT replied:
I fixed this in latest changes.

afuller-TT reviewed (changes requested):
Marking 'changes wanted' primarily for the circular dep, but also for tighter encapsulation of TT-Fabric.

Have only looked at the build structure of the library proper.

aliuTT commented:
@afuller-TT (I think) I've addressed all your feedback on CMakeLists, please take a look. Appreciate the feedback :)

aagarwalTT committed:
Fix tests

afuller-TT reviewed (commented):
Removing my "Request changes" status.  LMK if you need help addressing the clang-tidy error.

afuller-TT reviewed (approved):
Scratch that; the only way to remove it is to turn it into an actual Approve 🙃 
Still pending some updates; but I don't want to be blocking it once those are in.

aliuTT committed:
Point ERISC_MEM_BANK_TO_NOC_SCRATCH to ERISC_L1_UNRESERVED_BASE

aliuTT committed:
Add Fabric tests to CI

tt-rkim commented on .github/workflows/t3000-unit-tests-impl.yaml:
```diff
@@ -15,6 +15,7 @@ jobs:
       matrix:
         test-group: [
           { name: "t3k ttmetal tests", arch: wormhole_b0, cmd: run_t3000_ttmetal_tests, timeout: 30, owner_id: ULMEPM2MA}, #Sean Nijjar
+          { name: "t3k ttmetal tests", arch: wormhole_b0, cmd: run_t3000_ttfabric_tests, timeout: 30, owner_id: XXXXX}, # Add owner
```
Please add someone here - no `XXXXX`

aliuTT replied:
Ok, will add myself.

tt-rkim commented on CODEOWNERS:
```diff
@@ -47,6 +47,9 @@ tests/scripts/t3000/ @tenstorrent/metalium-developers-infra
 tests/scripts/tgg/ @tenstorrent/metalium-developers-infra
 
+# fabric
+tt_fabric/ @ubcheema @aliuTT
```
I would recommend another person, but at this point it won't be a big deal.

Until it becomes one.

aliuTT replied:
I will add Abhinav and Ellie who will contribute heavily to these files.

tt-rkim replied:
Who is Ellie?

aliuTT replied:
new hire working on Fabric features

tt-rkim commented:
@afuller-TT You can dismiss your review, by the way, since you're an admin :) 
<img width="862" alt="Screenshot 2025-01-10 at 4 51 03 PM" src="https://github.com/user-attachments/assets/1dfb8d53-f567-4e33-b251-a0a28299efd0" />

tt-rkim commented on tests/tt_metal/tt_fabric/common/n300_cluster_desc.yaml:
Are these descriptors new for `tt_fabric`? Where will these be used? And in which cluster object?

aliuTT replied:
They are needed for offline testing for the ControlPlane. Will not be used in silicon jobs.

omilyutin-tt commented on tt_fabric/routing_table_generator.cpp:
```diff
@@ -0,0 +1,249 @@
+    ss << "Routing Table Generator: IntraMesh Routing Tables" << std::endl;
+    for (mesh_id_t mesh_id = 0; mesh_id < this->intra_mesh_table_.size(); mesh_id++) {
+        ss << "M" << mesh_id << ":" << std::endl;
+        for (chip_id_t src_chip_id = 0; src_chip_id < this->intra_mesh_table_[mesh_id].size(); src_chip_id++) {
```
An alias like `auto& chip_routing = intra_mesh_table_[mesh_id];` (naming up to you) would makes sense here. Also in other places, where appropriate.

omilyutin-tt commented on tt_fabric/routing_table_generator.cpp:
```diff
@@ -0,0 +1,249 @@
+    }
+    log_debug(tt::LogFabric, "{}", ss.str());
+    ss.str(std::string());
+    ss << "Routing Table Generator: InterMesh Routing Tables" << std::endl;
+    for (mesh_id_t mesh_id = 0; mesh_id < this->inter_mesh_table_.size(); mesh_id++) {
+        ss << "M" << mesh_id << ":" << std::endl;
+        for (chip_id_t src_chip_id = 0; src_chip_id < this->inter_mesh_table_[mesh_id].size(); src_chip_id++) {
+            ss << "   D" << src_chip_id << ": ";
+            for (chip_id_t dst_chip_or_mesh_id = 0;
+                 dst_chip_or_mesh_id < this->inter_mesh_table_[mesh_id][src_chip_id].size();
+                 dst_chip_or_mesh_id++) {
+                auto direction = this->inter_mesh_table_[mesh_id][src_chip_id][dst_chip_or_mesh_id];
+                ss << dst_chip_or_mesh_id << "(" << magic_enum::enum_name(direction) << ") ";
+            }
+            ss << std::endl;
+        }
```
This is duplicated? Would a lambda that accepts either `inter_mesh_table_` or `intra_mesh_table_` work here?

omilyutin-tt commented on tt_fabric/mesh_graph.cpp:
```diff
@@ -0,0 +1,306 @@
+    }
+    log_debug(tt::LogFabric, "{}", ss.str());
+    ss.str(std::string());
+    ss << " Mesh Graph:  Inter Mesh Connectivity: " << std::endl;
+    for (uint32_t mesh_id = 0; mesh_id < this->inter_mesh_connectivity_.size(); mesh_id++) {
+        ss << "M" << mesh_id << ":" << std::endl;
+        for (uint32_t chip_id = 0; chip_id < this->inter_mesh_connectivity_[mesh_id].size(); chip_id++) {
+            ss << "   D" << chip_id << ": ";
+            for (auto [connected_mesh_id, edge] : this->inter_mesh_connectivity_[mesh_id][chip_id]) {
+                for (int i = 0; i < edge.connected_chip_ids.size(); i++) {
+                    ss << "M" << connected_mesh_id << "D" << edge.connected_chip_ids[i] << "("
+                       << magic_enum::enum_name(edge.port_direction) << ", " << edge.weight << ") ";
+                }
+            }
+            ss << std::endl;
+        }
+    }
```
Similar comment about code duplication, I think you can use a generic lambda here to handle both cases cleanly.

omilyutin-tt commented on tt_fabric/routing_table_generator.hpp:
```diff
@@ -0,0 +1,60 @@
+
+namespace tt::tt_fabric {
+
+using RoutingTable =
+    std::vector<std::vector<std::vector<RoutingDirection>>>;  // [mesh_id][chip_id][target_chip_or_mesh_id]
```
The printing functions can also go here.

Ideally we would also use something similar for `directions_` from within the wrapper class, but I think it is ok to construct the data member and set it externally once, but optimize for the nice type safe access API.

omilyutin-tt reviewed (approved):
Approve as per the offline discussion that the comments will be resolved in the follow up PR.
